src/query/sql/src/planner/semantic/type_check.rs

```diff
@@ -3919,6 +3966,82 @@ impl<'a> TypeChecker<'a> {
         }
     }
 
+    fn rewritable_variant_functions() -> &'static [Ascii<&'static str>] {
+        static VARIANT_FUNCTIONS: &[Ascii<&'static str>] = &[
+            Ascii::new("get"),
+            Ascii::new("get_string"),
+            Ascii::new("get_by_keypath"),
+            Ascii::new("get_by_keypath_string"),
+        ];
+        VARIANT_FUNCTIONS
+    }
+
+    fn try_rewrite_variant_function(
+        &mut self,
+        span: Span,
+        func_name: &str,
+        args: &mut [ScalarExpr],
+        arg_types: &mut [DataType],
+    ) -> Result<Option<Box<(ScalarExpr, DataType)>>> {
+        if !Self::rewritable_variant_functions().contains(&Ascii::new(func_name))
+            || arg_types.is_empty()
+            || arg_types[0].remove_nullable() != DataType::Variant
+        {
+            return Ok(None);
+        }
+        let ScalarExpr::BoundColumnRef(column_ref) = &args[0] else {
+            return Ok(None);
+        };
+        // only rewrite when arg[1] is path
+        let ScalarExpr::ConstantExpr(ConstantExpr {
+            value: Scalar::String(path),
```

Comment:
the second argument of `get` may be int, which is used for get variant array values.